src/java/org/apache/cassandra/index/sai/memory/MultiBlockIndex.java

```diff
@@ -0,0 +1,165 @@
+ * order until a match is found.
+ * @see BlockIndex
+ *
+ * Implements:
+ * @see MemoryIndex
+ */
+public class MultiBlockIndex extends MemoryIndex
+{
+    public static final int BLOCK_SIZE = 4000;
+    private final Object writeLock = new Object();
+    protected final ConcurrentLinkedQueue<BlockIndex> existingBlocks = new ConcurrentLinkedQueue<>();
+    protected BlockIndex current;
+
+    public MultiBlockIndex(IndexContext indexContext)
+    {
+        super(indexContext);
+    }
+
+    /**
+     * Add a key and value to the current block index.
+     * @param key Partition key
+     * @param clustering Clustering key
+     * @param value Index value
+     * @return Bytes memory used
+     */
+    @Override
+    public long add(DecoratedKey key, Clustering clustering, ByteBuffer value)
+    {
+        synchronized (writeLock)
+        {
+            if (current == null)
+                current = new BlockIndex(indexContext, this);
+
+            final long bytesUsed = current.add(key, clustering, value);
+
+            if (current.count() == BLOCK_SIZE)
+            {
+                existingBlocks.add(current);
+                current = null;
+            }
+            return bytesUsed;
+        }
+    }
+
+    /**
+     * Search the block indexes.
+     * @param expression Query
+     * @param keyRange Key range
+     * @return Matches range iterator
+     */
+    @Override
+    public RangeIterator search(Expression expression, AbstractBounds<PartitionPosition> keyRange)
+    {
+        final BlockIndex currentIndex = current;
+
+        final List<RangeIterator> iterators = new ArrayList<>();
+
+        if (currentIndex != null && currentIndex.contains(expression))
+        {
+            RangeIterator iterator = currentIndex.search(expression, keyRange);
+            if (iterator != null)
+                iterators.add(iterator);
+        }
+
+        for (BlockIndex blockIndex : existingBlocks)
+        {
+            if (blockIndex.contains(expression))
+            {
+                RangeIterator iterator = blockIndex.search(expression, keyRange);
+                if (iterator != null)
+                    iterators.add(iterator);
+            }
+        }
+
+        if (iterators.isEmpty())
+            return RangeIterator.empty();
+
+        return RangeUnionIterator.build(iterators);
+    }
+
+    /**
+     * @return Sorted iterator over index.
+     */
+    @Override
+    public Iterator<Pair<ByteComparable, PrimaryKeys>> iterator()
+    {
+        List<BlockIndex.ValueAndKey> list = new ArrayList<>();
+
+        if (current != null)
+        {
+            existingBlocks.add(current);
+            current = null;
+        }
+
+        for (BlockIndex blockIndex : existingBlocks)
+            blockIndex.gather(list);
+
+        Collections.sort(list);
+
+        ByteComparable lastValue = null;
+        PrimaryKeys keys = null;
+        List<Pair<ByteComparable, PrimaryKeys>> finalList = new ArrayList<>();
```

Comment:
> I believe this could be avoided.

Interesting.

Reply:
The `TrieMemoryIndex` avoids that. But let's leave it for a while, I'm not saying it would be easy.

The core of my comment is why do we think it is ok to duplicate the index data in memory when iterating? 
Such behaviour is not something that one would expect when calling a method that returns an iterator, so let's at least add a "heads up" note saying this method is potentially very costly.